subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -17,119 +17,141 @@
+    public CurrentSnapshotResult execute(IncrementalChangesContext context) {
         if (!buildCache.isEnabled()) {
             return executeWithoutCache(context);
         }
-        CacheHandler cacheHandler = context.getWork().createCacheHandler();
+        UnitOfWork work = context.getWork();
+        CacheHandler cacheHandler = work.createCacheHandler();
         return cacheHandler
-            .load(cacheKey -> load(context.getWork(), cacheKey))
-            .map(loadResult -> {
-                OriginMetadata originMetadata = loadResult.getOriginMetadata();
-                ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = loadResult.getResultingSnapshots();
-                return (CurrentSnapshotResult) new CurrentSnapshotResult() {
+            .load(cacheKey -> Try.ofFailable(() -> buildCache.load(commandFactory.createLoad(cacheKey, work))))
+            .map(successfulLoad -> successfulLoad
+                .map(cacheHit -> {
+                    OriginMetadata originMetadata = cacheHit.getOriginMetadata();
+                    ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = cacheHit.getResultingSnapshots();
+                    return (CurrentSnapshotResult) new CurrentSnapshotResult() {
+                        @Override
+                        public Try<ExecutionOutcome> getOutcome() {
+                            return Try.successful(ExecutionOutcome.FROM_CACHE);
+                        }
+
+                        @Override
+                        public OriginMetadata getOriginMetadata() {
+                            return originMetadata;
+                        }
+
+                        @Override
+                        public boolean isReused() {
+                            return true;
+                        }
+
+                        @Override
+                        public ImmutableSortedMap<String, CurrentFileCollectionFingerprint> getFinalOutputs() {
+                            return finalOutputs;
+                        }
+                    };
+                })
+                .orElseGet(() -> executeAndStoreInCache(cacheHandler, context))
+            )
+            .orElseMapFailure(loadFailure -> {
+                LOGGER.warn("Failed to load cache entry for {}, cleaning outputs and falling back to (non-incremental) execution",
+                    work.getDisplayName(), loadFailure);
+                cleanupTreesAfterLoadFailure(work);
+
+                return executeAndStoreInCache(cacheHandler, new IncrementalChangesContext() {
                     @Override
-                    public Try<ExecutionOutcome> getOutcome() {
-                        return Try.successful(ExecutionOutcome.FROM_CACHE);
+                    public Optional<ExecutionStateChanges> getChanges() {
+                        // Clear change information to avoid incremental execution after failed load
+                        return Optional.empty();
```

Comment:
❌ Doesn't clearing the changes mean something else? I thought it meant we don't track the inputs.
I think we need to return `RebuildExecutionStateChanges` here.

An incremental task would fail when the changes are empty.

Reply:
Ah, damn, fixed.